docs/articles/semantic_chunking.md

```diff
@@ -0,0 +1,537 @@
+        self.extractor = ExtractSentencesProgram()
+
+    def get_propositions(self, text):
+        sentences = self.extractor.run(text)
+
+        if isinstance(sentences, list):
+            return sentences
+
+        # Fallback: extract sentences heuristically
+        return [s.strip() for s in text.split('.') if s.strip()]
+
+    def split_text(self, text: str) -> List[str]:
+        """Extract propositions and chunk them accordingly."""
+        propositions = self.get_propositions(text)
+        return self._chunk_propositions(propositions)
+
+    def split_documents(self, documents: List[Document]) -> List[Document]:
+        """Split documents into chunks."""
+        split_docs = []
+        for doc in documents:
+            chunks = self.split_text(doc.page_content)
+            for i, chunk in enumerate(chunks):
+                metadata = doc.metadata.copy()
+                metadata.update({"chunk_index": i})
+                split_docs.append(Document(page_content=chunk, metadata=metadata))
+        return split_docs
+
+    def _chunk_propositions(self, propositions: List[str]) -> List[str]:
+        chunks = []
+        current_chunk = []
+        current_size = 0
+
+        for prop in propositions:
+            prop_size = len(prop)
+            if current_size + prop_size > self.chunk_size and current_chunk:
+                chunks.append(" ".join(current_chunk))
+                current_chunk = []
+                current_size = 0
+            current_chunk.append(prop)
+            current_size += prop_size
+
+        if current_chunk:
+            chunks.append(" ".join(current_chunk))
+
+        return chunks
+```
+
+## Evaluation
+
+We conducted experiments with the above three variations of semantic chunking on two datasets and embedding models.
+...to see how they performed in terms of...
+
+***Datasets***
+
+1. [HotpotQA](https://huggingface.co/datasets/hotpotqa/hotpot_qa?row=16)
+2. [SQUAD](https://huggingface.co/datasets/squad?row=0)
+
+***Embedding models***
+
+1. [BAAI/bge-small-en-v1.5](https://huggingface.co/BAAI/bge-small-en-v1.5)
+2. BAAI/bge-small-en-v1.5 with reranker
+
+Here are the results.
+
+![Evaluation results](../assets/use_cases/semantic_chunking/table.png)
+
+![Results summary](../assets/use_cases/semantic_chunking/graph5.png)
+
+From the above results, we can summarize.
+
+- LLM based semantic chunking performs best overall, especially in faithfulness and handling complex queries as LLMs can understand context and semantics at a deep level, allowing for more intelligent chunking. But it has high latency numbers as LLM inference takes time.
+- Embedding similarity based semantic chunking shows strong performance in relevancy and a good balance of precision and recall.
+- Clustering based semantic chunking is more variable, potentially excelling in specific scenarios but possibly struggling with others. It has slightly higher latency numbers compared to embedding based method.

```

Comment:
@Ashish-Abraham This is a bit vague. Can you add any more detail here?

Reply:
- LLM-semantic chunking emerged as the top performer overall, achieving the highest scores across most metrics. It particularly excelled in faithfulness, scoring 0.83 on hotpot_qa and 0.95 on squad without a reranker, increasing to 0.87 and 0.98 respectively with a reranker. This method also demonstrated strong performance on complex, multi-hop reasoning tasks represented by the hotpot_qa dataset, with relevancy scores of 0.86 (without reranker) and 0.88 (with reranker). This suggests that LLMs can better understand and preserve complex semantic relationships when chunking text.

- Embedding similarity based semantic chunking  stood out for its good relevancy scores across both datasets. It achieved scores of 0.89 on hotpot_qa and 0.93 on squad without a reranker, improving to 0.89 and 0.94 respectively with a reranker. This method also demonstrated a well-balanced precision and recall, with scores closely matching each other (precision: 0.77, recall: 0.79 on hotpot_qa without reranker).


- Cluster-semantic chunking performing decently but lower. It consistently scored lower across metrics compared to other methods (e.g., faithfulness: 0.63 for hotpot_qa, relevancy: 0.83 for squad without reranker). It's a viable method but may struggle with capturing fine-grained semantic relationships.

- The addition of a reranker consistently improved performance across all chunking methods. For example, LLM-semantic chunking on the squad dataset saw its relevancy increase from 0.92 to 0.94, and its precision from 0.94 to 0.97 with the addition of a reranker. 

- Latency considerations revealed that LLM-semantic chunking, while performing best, also had the highest latency (6.88 seconds for hotpot_qa, 5.83 seconds for squad). Other methods, such as semantic chunking, demonstrated lower latency (5.24 seconds on hotpot_qa), highlighting a trade-off between performance and speed.

Reply:
thanks, I've blended your two summaries.